src/python/pants/fs/archive.py

```diff
@@ -22,8 +22,30 @@
 class Archiver(AbstractClass):
 
   @classmethod
-  def extract(cls, path, outdir):
-    """Extracts an archive's contents to the specified outdir."""
+  def extract(cls, path, outdir, filter_func=None, concurrency_safe=False):
+    """Extracts an archive's contents to the specified outdir with an optional filter.
+
+    :API: public
+
+    :param string path: path to the zipfile to extract from
+    :param string outdir: directory to extract files into
+    :param function filter_func: optional filter with the filename as the parameter.  Returns True
+      if the file should be extracted.  Note that filter_func is ignored for non-zip archives.
+    :param bool concurrency_safe: True to use concurrency safe method.  Concurrency safe extraction
+      will be performed on a temporary directory and the extacted directory will then be renamed
+      atomically to the outdir.  As a side effect, concurrency safe extraction will not allow
+      overlay of extracted contents onto an existing outdir.
+    """
+    if concurrency_safe:
+      with temporary_dir() as temp_dir:
+        cls._extract(path, temp_dir)
+        safe_concurrent_rename(temp_dir, outdir)
```

Comment:
Note that since the outdir passed in is the parent of the location of the downloaded `node.tar.gz` file, that file will disappear, leading to BinaryUtils downloading it again every time. Solution is probably to add `/unpacked` to the outdir, so it's a sibling of the tar.gz file.